### DISCUSSION TITLE
Gradients tracking in case of integer data type.

### DISCUSSION THREAD

**Discussion by sachinlodhi** [mrdbourke/pytorch-deep-learning#1106] [Q&A] [answered]
I have intialized a random tensor of int32 type and trying to set the requires_grad=True like this:
`python
int_32_tensor = torch.tensor([1,6,73], dtype = torch.int32, 
                            device= "cuda",
                            requires_grad = True)`
and this throws an error as:
`RuntimeError: Only Tensors of floating point and complex dtype can require gradients`
I get the idea that functions with integer values are not continuous, and calculating gradient for them is a kinda dumb thing to do.
But any other explanation?

**Comment by LuluW8071** [comment, marked as answer]
Yes, you are correct. Calculating gradients relies on calculating changes even if that in extremely small during backpropagation. 

In case of integer, it is just discrete meaning there might be no huge jump from going to $1$ to $2$ and huge jump from going to $1$ to $10$. Such sudden changes in jump results to non-smooth and non-converging behaviour.

But in case of floating values as you said its continuous, even infinitely small changes in its input helps during backprop thats whyy gradient calculation is limited to float and complex types.
